torchgeo/datasets/stacapidataset.py

```diff
@@ -0,0 +1,321 @@
+from torchgeo.datasets.utils import BoundingBox
+
+# from torchgeo.samplers import RandomGeoSampler
+
+
+class STACAPIDataset(GeoDataset):
+    """STACApiDataset.
+
+    SpatioTemporal Asset Catalogs (`STACs <https://stacspec.org/>`_) are a way
+    to organize geospatial datasets. STAC APIs let you query huge STAC Catalogs by
+    date, time, and other metadata.
+
+
+    .. versionadded:: 0.3
+    """
+
+    sentinel_bands = [
+        "B01",
+        "B02",
+        "B03",
+        "B04",
+        "B05",
+        "B06",
+        "B07",
+        "B08",
+        "B8A",
+        "B09",
+        "B11",
+        "B12",
+    ]
+
+    def __init__(  # type: ignore[no-untyped-def]
+        self,
+        root: str,
+        crs: Optional[CRS] = None,
+        res: Optional[float] = None,
+        bands: Sequence[str] = sentinel_bands,
+        is_image: bool = True,
+        api_endpoint: str = "https://planetarycomputer.microsoft.com/api/stac/v1",
+        transforms: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+        **query_parameters,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            crs: :term:`coordinate reference system (CRS)` to warp to
+                (defaults to the CRS of the first file found)
+            res: resolution of the dataset in units of CRS
+                (defaults to the resolution of the first file found)
+            bands: sequence of of stac asset band names
+            is_image: if true, :meth:`__getitem__` uses `image` as sample key, `mask`
+                otherwise
+            api_endpoint: api for pystac Client to access
+            transforms: a function/transform that takes an input sample
+                and returns a transformed versio
+            query_parameters: parameters for the catalog to search, for an idea see
+            <https://pystac-client.readthedocs.io/en/latest/api.html#pystac_client.ItemSearch>
+        """
+        self.root = root
+        self.api_endpoint = api_endpoint
+        self.bands = bands
+        self.is_image = is_image
+
+        # Create an R-tree to index the dataset
+        self.index = Index(interleaved=False, properties=Property(dimension=3))
+
+        catalog = Client.open(api_endpoint)
+
+        search = catalog.search(**query_parameters)
+
+        items = list(search.get_items())
+
+        if not items:
+            raise RuntimeError(
+                f"Your search criteria off {query_parameters} did not return any items"
+            )
+
+        epsg = items[0].properties["proj:epsg"]
+        crs_dict = {"init": "epsg:{}".format(epsg)}
+        src_crs = CRS.from_dict(crs_dict)
+        if crs is None:
+            crs = CRS.from_dict(crs_dict)
+
+        for i, item in enumerate(items):
+            minx, miny, maxx, maxy = item.bbox
+
+            transformer = Transformer.from_crs(src_crs.to_epsg(), crs.to_epsg())
```

Comment:
The output from `item.bbox` appears to be longitude/latitude coordinates, i.e. EPSG:4326. I don't know if this is the case for **all** STAC datasets, but assuming that `item.bbox` is always EPSG:4326, then the pyproj.Transformer should be doing the coordinate transform like so. Oh, and usually a good idea to use `always_xy=True`.

```suggestion
            minx, miny, maxx, maxy = item.bbox

            transformer = Transformer.from_crs(4326, crs.to_epsg(), always_xy=True)
```